eth/gasprice/gasprice.go

```diff
@@ -17,212 +17,152 @@
+	GpoBlocks     int
+	GpoPercentile int
+	GpoDefault    *big.Int
 }
 
 // GasPriceOracle recommends gas prices based on the content of recent
-// blocks.
+// blocks. Suitable for both light and full clients.
 type GasPriceOracle struct {
-	chain         *core.BlockChain
-	db            ethdb.Database
-	evmux         *event.TypeMux
-	params        *GpoParams
-	initOnce      sync.Once
-	minPrice      *big.Int
-	lastBaseMutex sync.Mutex
-	lastBase      *big.Int
-
-	// state of listenLoop
-	blocks                        map[uint64]*blockPriceInfo
-	firstProcessed, lastProcessed uint64
-	minBase                       *big.Int
+	backend   ethapi.Backend
+	lastHead  common.Hash
+	lastPrice *big.Int
+	cacheLock sync.RWMutex
+	fetchLock sync.Mutex
+
+	checkBlocks, minBlocks, maxBlocks int
+	percentile                        int
 }
 
 // NewGasPriceOracle returns a new oracle.
-func NewGasPriceOracle(chain *core.BlockChain, db ethdb.Database, evmux *event.TypeMux, params *GpoParams) *GasPriceOracle {
-	minprice := params.GpoMinGasPrice
-	if minprice == nil {
-		minprice = big.NewInt(gpoDefaultMinGasPrice)
+func NewGasPriceOracle(backend ethapi.Backend, params GpoParams) *GasPriceOracle {
+	blocks := params.GpoBlocks
+	if blocks < 1 {
+		blocks = 1
+	}
+	percent := params.GpoPercentile
+	if percent < 0 {
+		percent = 0
 	}
-	minbase := new(big.Int).Mul(minprice, big.NewInt(100))
-	if params.GpobaseCorrectionFactor > 0 {
-		minbase = minbase.Div(minbase, big.NewInt(int64(params.GpobaseCorrectionFactor)))
+	if percent > 100 {
+		percent = 100
 	}
 	return &GasPriceOracle{
-		chain:    chain,
-		db:       db,
-		evmux:    evmux,
-		params:   params,
-		blocks:   make(map[uint64]*blockPriceInfo),
-		minBase:  minbase,
-		minPrice: minprice,
-		lastBase: minprice,
+		backend:     backend,
+		lastPrice:   params.GpoDefault,
+		checkBlocks: blocks,
+		minBlocks:   (blocks + 1) / 2,
+		maxBlocks:   blocks * 5,
+		percentile:  percent,
 	}
 }
 
-func (gpo *GasPriceOracle) init() {
-	gpo.initOnce.Do(func() {
-		gpo.processPastBlocks()
-		go gpo.listenLoop()
-	})
-}
-
-func (self *GasPriceOracle) processPastBlocks() {
-	last := int64(-1)
-	cblock := self.chain.CurrentBlock()
-	if cblock != nil {
-		last = int64(cblock.NumberU64())
-	}
-	first := int64(0)
-	if last > gpoProcessPastBlocks {
-		first = last - gpoProcessPastBlocks
-	}
-	self.firstProcessed = uint64(first)
-	for i := first; i <= last; i++ {
-		block := self.chain.GetBlockByNumber(uint64(i))
-		if block != nil {
-			self.processBlock(block)
+// SuggestPrice returns the recommended gas price.
+func (self *GasPriceOracle) SuggestPrice(ctx context.Context) (*big.Int, error) {
+	self.cacheLock.RLock()
+	lastHead := self.lastHead
+	lastPrice := self.lastPrice
+	self.cacheLock.RUnlock()
+
+	head, _ := self.backend.HeaderByNumber(ctx, rpc.LatestBlockNumber)
+	headHash := head.Hash()
+	if headHash == lastHead {
+		return lastPrice, nil
+	}
+
+	self.fetchLock.Lock()
+	defer self.fetchLock.Unlock()
```

Comment:
I'm a bit unsure about this pattern here. Fetching the last N blocks may run for a considerable amount of time. The current approach seems to flat out block on it. That's ok for a full client since fetching the blocks should be fast. But how much will a light client block? Perhaps a saner thing would be to try and update the gas price in the background, but return the previous price to the user if X seconds pass?